### DISCUSSION TITLE
Undercut, oval toneholes, and a fingering optimizer

### DISCUSSION THREAD

**Discussion by chihowkwan** [edwardkort/WWIDesigner#141] [Ideas]
Hello there! Thank you for creating WIDesigner, and I use it to design and make transverse flutes. The kind of flutes I made relates to the general idea of Chinese dizi, and I hope WIDesigner might add some features related to my research in the future.

The general design of most Chinese dizi includes toneholes with an oval shape, and usually some of them are heavily undercut. I wonder whether the undercut could be counted by WIDesigner, and whether a function of deciding the proportion of ovalness of toneholes could be added. Comparing oval to circular toneholes, players of the dizi not only prefer the allegedly enhancement of the timber, but also the more comfort and safety the oval geometry causes to the hole-closing fingers. 

Sometimes I would analyze possibilities for my transverse flutes to go up to their third octave. The toneholes I have are 11. As it is a common feature that fingerings for the third octave usually become strange, I can’t determine whether a note did not sound is due to the limitation of the design or a mistake in the fingering. Currently I try different fingerings in a real instrument and type in the result for further analysis. However, I wonder whether there can be a fingering optimizer which finds a best fingering for me for a note in WIDesigner.

Thank you for your attention!

**Comment by bhp1** [comment]
Unfortunately, for oval toneholes and undercutting we don't have a mathematical model of how the change in tonehole shape affects the acoustics.  Antoine Lefebvre developed the tonehole model we use through extensive finite-element modelling.  It would take a similar project, or extensive experimentation, to adapt the model to oval toneholes or undercutting.  We would need to quantify how oval the tonehole was, or how much undercutting was done, and include that in the model.

A fingering finder, on the other hand, might be feasible.  For a static geometry, it wouldn't take WIDesigner long to try all 2048 fingerings to find the ones that came close to a desired frequency.  I'd have to give some thought on how to incorporate it into the user interface and program structure.

**Reply by chihowkwan** [reply]
Thank you for your kind reply. I’m very glad to not have to click 2048 times for a result! Looking forward to the new function, I would like to mention a feature common in Chinese dizi which might need some kind of constraints of the fingering finder. A hole covered with a membrane is usually seen next to the embouchure hole and the membrane vibrates when playing, giving the flute a unique timbre. In designing I set this hole to be always closed (actually by the membrane) in the fingering file, and there is no need to open it. I do hope the fingering finder can let me keep the holes I choose to stay the same when it is going to calculate fingerings.

**Comment by bhp1** [comment]
Have you found that the model gives good results when you treat the membrane hole as an always-closed hole?  Does it give better results than if you leave the hole out of the model entirely?  I would suspect that in real life, because of the membrane, that hole wouldn't look quite like a closed hole.  In addition, the model assumes the hole is closed with a finger pad that intrudes into the hole to some degree, rather than a flat closure at the outside end of the hole.  I'm not sure there's a good way to represent this in WIDesigner at present.

**Comment by bhp1** [comment]
If we do implement a fingering finder, I don't think we need to constrain what holes it tests.

The fingering finder needs to return not just one candidate fingering, but as many as it can find.  There are cases where a fingering will show a resonance at a particular frequency, but in real life you can't play that fingering at that frequency.  In [this example](https://github.com/edwardkort/WWIDesigner/wiki/Reading-the-Note-Spectrum-Graph#what-the-graph-tells-us), the fingering OXXXXX has a resonance around 1950 Hz, but the whistle can't actually produce that note with that fingering.  The fingering finder for 1950 Hz would give OXXXXX as one of the candidates.  You would have to try out that candidate and discard it in favour of another candidate that does work, XOOOOO.

Similarly, if one of the candidates opens a hole you want always-closed, or closes a hole you want always-open, you can just ignore that candidate and look at others.

**Comment by edwardkort** [comment]
I'm curious to hear real-world use cases for a fingering finder. 

When designing an instrument, whether it be an Irish whistle, a baroque flute, a recorder, a native American-style flute (NAF), or a dizi, there is typically a "standard" fingering set (or a few sets). This is expected by any player who will use this instrument.

Yes, there are alternate fingerings for many notes - to be used for quick passages, trills, and other ornamentations. But such fingerings are not expected to be in perfect tune, and so are not included in the design calculations.

In designing NAFs, I have only two use cases for investigating alternate fingering sets (what we call tunings in WIDesigner):
1. There is one note that I want to be in tune using two fingerings. Depending on the flute, this is not always possible, and this can drive the tuning of several other notes in an adverse fashion. So, I often must manually evaluate the four fingering possibilities (including not using this note) manually.
2. The NAF, at its best, is at most a 15-note chromatic instrument; at its worst, only a 6-note pentatonic one. For difficult designs, I must sequentially add notes to the pentatonic base until the overall tuning is unacceptable.

Neither of the above use cases would be aided by an unconstrained fingering finder.

So again, I would love to see the use cases that would make this programming effort worthwhile.

Cheers,
Edward Kort
Wood Wind

**Reply by chihowkwan** [reply]
Thank you Mr. Kort for your thoughtful reply. I would happily share with you the situation in which I might need a fingering finder, and I’m so grateful to see developers of WIDesigner discuss issues with me. I’m currently using WIDesigner to analyze and optimize (to a certain extent) a kind of Chinese dizi which may never get popularized even in China or Taiwan. It is called xindi (literally “new dizi”) by its inventor Ding, Xi-Lin in 1935. Ding studied physics in University of Birmingham and received a master degree in 1919. He came back to China and invented xindi when he was the director in the physics institute of China’s Academia Sinica as part of his research results. No evidence showed that he has continued improving this instrument since then.

You can regard xindi as a product influenced by the Boehm flute. Ding pioneered his people for seeing a need for dizi to transpose to any key in music which Boehm did a great job. He then chose to follow the equal temperament and design 11-hole dizi, xindi, one hole for one note. No keys were added.

A few months ago I found an old book which contains Ding’s published article in 1935 and had it shipped from China. Fortunately he provided a relatively full data (obviously he did not include parameters related to the performer) of one of his xindi design in this article. The following is the part of that data: 

![IMG_5774](https://user-images.githubusercontent.com/85783979/145073920-d132c541-9c71-47be-9d85-4eb6a43c1ca0.JPG)


Ding did not include a fingering chart in his article. The earliest and probably the only fingering chart for xindi was found in a book first published in 1962 written by a musician Gao, Zi-Ming who performed and composed for xindi since its birth. The book was intended to introduce xindi and its music to the public. It has a fingering chart which I think is developed mainly by Gao. The following fingering chart is from Gao’s book:

<img width="1591" alt="截圖 2021-12-08 上午1 06 00" src="https://user-images.githubusercontent.com/85783979/145074155-91eeb31b-9a6b-45a0-b358-498bb50c49e3.png">

As far as I know, no people researched into this chart or Ding’s xindi data. Ding actually hopes in the article that this data would be used for manufacturing and further researching. I think maybe this subject is too unpopular for people to research or maybe we do not have proper acoustic tools to analyze it. However, I realized that WIDesigner provides a way for people like me with no background on acoustics and programming to be able to control important aspects of woodwinds and see how they interact. I hope to analyze and optimize (to a certain extent) Ding’s xindi data and write the whole process down as my thesis for a master’s degree in ethnomusicology. I expect this thesis to be insightful for xindi design and hope to introduce WIDesigner to people who are interested in dizi designing. I hope the process would force me to learn more about musical acoustics as well.

The following screenshot was the result of the combination of Ding’s data, Gao’s fingering chart, and my rough calibration. Setting A4 as 435 Hz was the decision after observing even more tragical results using A4 = 440 Hz and knowing the 435 was historically possible in 1935. About the calibration, until now I have only made a required tube with an required embouchure hole to produce D4 and its harmonics and using them to calibrate. In the future I hope to acquire the complete instrument using light-cured 3D printing. Sorry this file still contains the membrane hole which I now realize is not appropriate. Apart from possible defects in analysis, from a designer’s perspective this instrument is not quite in tune. I especially don’t know what happened to A6 and A#6, therefore I plan to try possible fingerings for these notes on WIDesigner to see if it helps to improve a little bit. I also plan to try them out on the instrument printed. This is the intention I have for a fingering finder. Optimizers and some manual operations will be discussed after using this method in my thesis.

<img width="2048" alt="截圖 2021-12-08 上午12 14 58" src="https://user-images.githubusercontent.com/85783979/145074215-d8493321-76a4-4d58-96e5-dc0d006638b9.png">

Before seeing the above data and fingering chart, a friend of mine and I have already cooperated to design and make our version of xindi. The following are two examples, one with a membrane hole (now without a membrane on it) and one without; the black one still needs further polishing, buffing, and threading. We used WIDesigner to redesign the 11-hole xindi. This is because a friend of my friend owns and plays a xindi, but his xindi is quite out of tune; he therefore suggested us to redesign and make the xindi. The three-part design, decorative rings, and direction holes of the footjoint are ideas we add to xindi.

![IMG_7169](https://user-images.githubusercontent.com/85783979/145074341-e32051e8-7988-40bb-be59-ed1aa1129b3a.jpeg)
![IMG_7177](https://user-images.githubusercontent.com/85783979/145074363-f2857d84-173b-4684-a67c-7cd2f35d7427.jpeg)

We developed our version of fingering for our version of xindi. Knowing WIDesigner uses a linear blowing model and dizi family sounds out of tune in most of its third octave, we only gave D4-E6 weights while optimizing. In reality we gave up designing xindi for notes higher than E6. Notes higher than E6 are typed into the file only for appreciation, with zero weight. I do not plan to discuss our version of xindi in my thesis.

I therefore gain an understanding after this process of replying. For researching into other people’s design, fingering finder might be useful if a given fingering seems to produce weird result. In making one’s own design, fingering finder might be needed only when one still intend one note to be incorporated into the design.

Sorry for writing so much! It will take me another day or two to reply to other posts. I feel really blessed from reading your thoughts.
Sincerely,
Allen Kwan

**Reply by bhp1** [reply]
I agree, Edward, that a fingering finder would be a niche function.  However, if it isn't too difficult to implement, it is a lot easier for WIDesigner to check the 2^n fingerings for likely candidates than for the user, so it could save a very few users quite a bit of work.

In a way, a fingering finder, which finds fingerings to match a frequency, is the inverse of the note spectrum analyzer, which finds frequencies to match a fingering.

**Reply by edwardkort** [reply]
Burton, either you are more optimistic than me (most people are) or you have thought of a clever algorithm for implementing the fingering finder.

I envision many of the enumerated fingerings resulting in a tone that doesn't voice even though it has an in-tune harmonic. In note spectrum analyzer parlance, one of the surrounding harmonics has a higher loop gain. And BTW, I implemented a method/class in that analyzer framework that finds the loop gain peaks and prints their frequency and gain to the console when running the spectrum (I don't remember whether it only prints in the NAF study model). Wouldn't such functionality need to be incorporated into the optimizer's objective function? And how (it's an apple among oranges)?

The second challenge is arranging all the enumerated tuning results in some understandable order, with sufficient information to the user for evaluation.

**Reply by bhp1** [reply]
At this point, Edward, it is blind optimism; I don't have a clever algorithm, at least not yet.  At this point, I envision a brute force search of all 2^n possibilities, rather than an optimization, not unlike the loop gain maxima finder in PlayingRangeSpectrum, displaying any candidate fingerings that come close as they are found.  Yes, as least some of the candidates found wouldn't voice in practice, 

I may end up abandoning the idea when I get around to working on it, which may not be for a few weeks.

**Reply by chihowkwan** [reply]
Hello! I just read an article which might relate to the concept of the fingering finder: "Using acoustic modelling to design and print a microtonal clarinet" by Nicholas J Bailey, Théo Cremel, and Alex South. It mentions an algorithm DEAP (Distributed Evolutionary Algorithms in Python) which is "efficient in finding the best fingering set over a given instrument geometry". The instrument in this research is a clarinet with a cylindrical bore and nine toneholes, and the aim is to make it play in 19-EDO. So the algorithm is used to find proper fingerings for playing in 19-EDO. I would like to share with you and understand a little more about it!

**Reply by bhp1** [reply]
An interesting article, but not directly applicable to WIDesigner.  For modelling the clarinet, they use a time-domain simulation, whereas WIDesigner uses frequency domain modelling.  To identify candidate fingerings, they use an exhaustive search of all 2^9 possible fingerings on their 9-hole instrument and return the best one for each note.  The DEAP algorithm is used to search for an instrument geometry for which these best fingerings best render the target scale.

If I ever get to it, my proposed fingering finder was going to use an exhaustive search as well, potentially returning more than one candidate fingering per note, since some of them may not sound in practice. Once you pick a set of fingerings, you can use WIDesigner's existing optimization capabilities to tweak the instrument geometry to improve the tuning.
